protocol/src/protocol.typeHierarchy.proposed.md

```diff
@@ -0,0 +1,195 @@
+#### Type Hierarchy
+
+The type hierarchy language feature provides type hierarchy information for languages supporting inheritance and nominal typing.
+
+_Client Capabilities_:
+
+If the `textDocument.typeHierarchy.typeHierarchy` is `true`, then the language client is capable of sending super- and subtype hierarchy requests to the language server, and can represent the type hierarchy information.
+
+```ts
+/**
+ * Client capabilities specific to the type hierarchy feature.
+ */
+export interface TypeHierarchyCapabilities {
+
+  /**
+   * The text document client capabilities.
+   */
+  textDocument?: {
+
+    /**
+     * Capabilities specific to the `textDocument/typeHierarchy`.
+     */
+    typeHierarchy?: {
+
+      /**
+       * `true` if the language client supports super- and subtype hierarchies. Defaults to `false`.
+       */
+      typeHierarchy?: boolean;
+
+    }
+
+  }
+}
+```
+
+_Server Capabilities_:
+
+When the `typeHierarchy` language server capability is set to `true`, then the language server can provider hierarchical information to the client.
+
+```ts
+/**
+ * Type hierarchy language server capability.
+ */
+export interface TypeHierarchyServerCapabilities {
+
+  /**
+   * Server capability for calculating super- and subtype hierarchies.
+   * The language server supports the type hierarchy language feature, if this capability is set to `true`.
+   */
+  typeHierarchy?: boolean;
+
+}
+```
+
+_Request_:
+
+The `Type Hierarchy Request` request is sent from the client to the server to retrieve a `TypeHierarchyItem` at a given cursor position. If computing full type hierarchy items is expensive, servers can additionally provide a handler for the type hierarchy item resolve request (`typeHierarchy/resolve`). This request would also allow to specify if the item should be resolved and whether sub- or supertypes are to be resolved.
+
+#### Type Hierarchy Request:
+
+ * method: `textDocument/typeHierarchy`
+ * params: `TypeHierarchyParams`
+
+```ts
+/**
+ * The type hierarchy params is an extension of the `TextDocumentPositionParams` with optional properties
+ * which can be used to eagerly resolve the item when retrieving it from the server.
+ */
+export interface TypeHierarchyParams extends TextDocumentPositionParams {
+
+  /**
+   * The hierarchy levels to resolve. `0` indicates no level.
+   */
+  resolve?: number;
+
+  /**
+   * The direction of the hierarchy levels to resolve. Valid values are: `parents`, `children`, or `both`.
+   * Defaults to `children`.
+   */
+  direction?: 'parents' | 'children' | 'both';
```

Comment:
I think we need to clarify what should happen if a hierarchy item has already the parents resolved and is asked to resolve its children. Should the client send the parents again or is it allowed to leave them out. I would opt for the second.

Reply:
I do like strings literals over numbers but so far we used numbers for all enum like types. To keep things consistent I would like to opt to say with this unless we have a strong reason to do otherwise.

Reply:
I know I proposed this earlier but now I am more for consistency :-)

Reply:
> I would opt for the second.

+1

@dbaeumer, I would have a follow-up question on this: what should happen in general, if something is already resolved and the client requests to resolve it again?

In the JDT reference implementation, I did for this language feature, I blindly resolve everything from scratch, because:
 - the text document that is referenced by the `TypeHierarchyItem` might have changed meanwhile, so the item I am trying to resolve is not at the same cursor position anymore,
 - the type hierarchy of my `TypeHierarchyItem` might have changed.

Thank you!